File: app/src/main/java/com/example/miniproject/Repository/SourceRepository.java
package com.example.miniproject.Repository;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.content.Context;
import android.util.Log;

import com.example.miniproject.Database.Controller.DatabaseController;
import com.example.miniproject.Model.AbstractResponse;
import com.example.miniproject.Model.SourceModel;
import com.example.miniproject.RetrofitService.RetrofitController;
import com.example.miniproject.RetrofitService.ServiceClient;
import com.example.miniproject.Utils.AppExecutors;

import java.util.Calendar;
import java.util.Date;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SourceRepository {

    private static final String TAG = "SourceRepository";
    private final Context context;

    public SourceRepository(Context context) {
        this.context = context;
    }

    public LiveData<List<SourceModel>> getSourceListFromDb() {
        LiveData<List<SourceModel>> data;
        data = DatabaseController.getInstance(context).sourceDao().loadAllSources();
        return data;
    }

    public void getSourceListFromWebAndInsertInDB() {
        ServiceClient serviceClient = RetrofitController.getInstance().create(ServiceClient.class);
        serviceClient.getSourcesList().enqueue(new Callback<AbstractResponse>() {
            @Override
            public void onResponse(Call<AbstractResponse> call, final Response<AbstractResponse> response) {
                if (response.body() != null && response.body().getStatus().equals("ok"))
                    AppExecutors.getInstance().getDiskIO().execute(new Runnable() {
                        @Override
                        public void run() {
                            Calendar date = Calendar.getInstance();
                            date.add(Calendar.HOUR_OF_DAY,-1);
                            DatabaseController.getInstance(context).sourceDao().deleteAllSources(date.getTime().getTime());
                            DatabaseController.getInstance(context).sourceDao().insertAllSources(response.body().getSources());
                        }
                    });
            }

            @Override
            public void onFailure(Call<AbstractResponse> call, Throwable t) {
                Log.d(TAG, "onFailure: " + t);
            }
        });
    }

}

File: app/src/main/java/com/example/miniproject/repository/SourceRepository.kt
package com.example.miniproject.repository

import com.example.miniproject.database.controller.DatabaseController
import com.example.miniproject.model.AbstractResponse
import com.example.miniproject.retrofitService.ServiceClient
import com.example.miniproject.utils.AppExecutors
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*
import javax.inject.Inject

class SourceRepository @Inject constructor(
    private val databaseController: DatabaseController,
    private val serviceClient: ServiceClient
) {

    fun getSourceListFromDb() = databaseController.sourceDao().loadAllSources()

    val sourceListFromWebAndInsertInDB: Unit
        get() {
            serviceClient.sourcesList?.enqueue(object : Callback<AbstractResponse?> {
                override fun onResponse(
                    call: Call<AbstractResponse?>,
                    response: Response<AbstractResponse?>
                ) {
                    if (response.body() != null && (response.body()?.status == "ok")) {
                        AppExecutors.instance?.diskIO?.execute {
                            val date = Calendar.getInstance()
                            date.add(Calendar.HOUR_OF_DAY, -1)
                            databaseController.sourceDao()
                                .deleteAllSources(date.time.time)
                            databaseController.sourceDao()
                                .insertAllSources(response.body()?.sources)
                        }
                    }
                }

                override fun onFailure(call: Call<AbstractResponse?>, t: Throwable) {
                    t.printStackTrace()
                }
            })
        }

}